# U.S. Medical Insurance Costs

In [1]:
#Importing csv library
import csv

In [2]:
#Empty lists fot storing data of insurance.csv

age = []
sex = []
bmi = []
children = []
smoker = []
region = []
charges = []

In [3]:
# Define a function to extract data from the dataset and import it into the lists.

def addToList(my_list,csv_file,column):
    with open (csv_file, 'r') as csv_file:
        reader = csv.DictReader(csv_file)
        for row in reader:
            my_list.append(row[column])
        


In [4]:
# Execute the function to add values from the dataset to the list.

addToList(age, 'Insurance.csv', "age")
addToList(sex, 'Insurance.csv', "sex")
addToList(bmi, 'Insurance.csv', "bmi")
addToList(children, 'Insurance.csv', "children")
addToList(smoker, 'Insurance.csv', "smoker")
addToList(region, 'Insurance.csv', "region")
addToList(charges, 'Insurance.csv', "charges")

* find average age of the patients
* return the number of males vs. females counted in the dataset
* find geographical location of the patients
* return the average yearly medical charges of the patients
* return the median yearly medical charges of the patients
* creating a dictionary that contains all patient information
* Smokers per area

In [5]:
class Patients:
    def __init__(self, patients_age, patients_sex, patients_bmi, patients_children, patients_smoker, patients_region, patients_charges):
        # Initialize the Patients class with patient information attributes
        self.patients_age = patients_age
        self.patients_sex = patients_sex
        self.patients_bmi = patients_bmi
        self.patients_children = patients_children
        self.patients_smoker = patients_smoker
        self.patients_region = patients_region
        self.patients_charges = patients_charges

    def __repr__(self):
        # Define the string representation of the Patients class
        return f"Patient Info: Age={self.patients_age}, Sex={self.patients_sex}, BMI={self.patients_bmi}, Children={self.patients_children}, Smoker={self.patients_smoker}, Region={self.patients_region}, Charges={self.patients_charges}"

    def AverageAge(self, patients_age):
        # Calculate the average age of patients
        averageAge = 0
        for age in patients_age:
            averageAge += float(age)
        return "The average age is {}".format(round(averageAge / len(patients_age)))

    def num_males_females(self, patients_sex):
        # Count the number of males and females in the list and calculate the difference
        malecount = patients_sex.count('male')
        femalecount = patients_sex.count('female')

        if malecount != femalecount:
            primary_gender = "males" if malecount > femalecount else "females"
            secondary_gender = "females" if primary_gender == "males" else "males"

            dif = abs(malecount - femalecount)
            percent_dif = (dif / (malecount + femalecount)) * 100
            percent_total = 100 - percent_dif

        return f"There are {malecount} {primary_gender} and {femalecount} {secondary_gender} in the list.\n" \
               f"The difference is {dif}, which represents {percent_dif:.2f}% of difference\n"

        # Note: The second return statement is unreachable and can be removed.

    def region(self, patients_region):
        # Count the occurrences of each region in the list
        region_dic = {}
        for region in patients_region:
            counter = 0
            if region not in region_dic:
                region_dic[region] = 1
            else:
                region_dic[region] += 1
        return region_dic

    def averageYearlyCharges(self, patients_charges):
        # Calculate the average yearly charges of patients
        total_charge = 0
        for charge in patients_charges:
            total_charge += float(charge)

        average_charge = total_charge / len(patients_charges)
        return "The average yearly charge is {:.2f}".format(average_charge)
    
    def medianYearlyCharges(self, patients_charges):
        charges_sorted = sorted(map(float, patients_charges))
        n = len(charges_sorted)
        if n % 2 == 1:
        # Si la longitud es impar, devolver el valor en la posición central
            median_value = charges_sorted[n // 2]
        else:
        # Si la longitud es par, devolver el promedio de los dos valores centrales
            middle1 = charges_sorted[(n // 2) - 1]
            middle2 = charges_sorted[n // 2]
            median_value = float((middle1 + middle2) / 2)

        return "The median yearly charge is {:.2f}".format(median_value)


    def createDictionary(self):
        # Create a dictionary with patient information
        self.patients_dictionary = {}
        self.patients_dictionary['age'] = [int(age) for age in self.patients_age]
        self.patients_dictionary['sex'] = self.patients_sex
        self.patients_dictionary['bmi'] = self.patients_bmi
        self.patients_dictionary['children'] = self.patients_children
        self.patients_dictionary['smoker'] = self.patients_smoker
        self.patients_dictionary['region'] = self.patients_region
        self.patients_dictionary['charges'] = self.patients_charges
        return self.patients_dictionary

    def smokers_by_region(self):
        # Count the number of smokers in each region
        smokers_by_region = {}
        for i, region in enumerate(self.patients_region):
            if self.patients_smoker[i] == 'yes':
                smokers_by_region[region] = smokers_by_region.get(region, 0) + 1
        return smokers_by_region

In [6]:
# Creating an instance of the Patients class with patient data
paciente = Patients(age, sex, bmi, children, smoker, region, charges)

In [7]:
# Calculate and print the average age of patients
print(paciente.AverageAge(age))

The average age is 39


Average Age:

The calculated average age of 39 suggests that the majority of patients in the dataset fall around this age. This measure provides a central tendency, but further exploration, such as examining age ranges or creating a histogram, would offer a more complete understanding of the age distribution.

In [8]:
# Count the number of males and females, and print the results
print(paciente.num_males_females(sex))

There are 676 males and 662 females in the list.
The difference is 14, which represents 1.05% of difference



Number of Males and Females:
The dataset exhibits a relatively balanced distribution between males (6766) and females (662). The small difference of 14 individuals, representing approximately 1.05% of the total, indicates a near-equitable gender representation.

In [9]:
# Count the occurrences of each region and print the results
print(paciente.region(region))

{'southwest': 325, 'southeast': 364, 'northwest': 325, 'northeast': 324}


Count of Patients in Each Region:
The regional distribution of patients reveals variations, with the Southeast having the highest count (364) and the Northeast the lowest (324). Understanding regional differences can be crucial for targeted healthcare interventions or resource allocation.

In [10]:
# Calculate and print the average yearly charges of patients
print(paciente.averageYearlyCharges(charges))

The average yearly charge is 13270.42


Average Yearly Charges:

The average yearly charge of $13,270.42 provides a central measure for healthcare costs. Further investigation into the distribution of charges, potential outliers, or a comparison with other factors like region or smoking status could offer more nuanced insights.

In [11]:
# Calculate and print the median yearly charges of patients

print(paciente.medianYearlyCharges(charges))


The median yearly charge is 9382.03


Median Yearly Charges:

The analysis suggests that the median yearly charge is $9382.03. The recommendation is to consider the median instead of the average (mean) charges when assessing the central tendency of the data. This is advised because outliers in the dataset can significantly impact the mean, causing it to be skewed. By focusing on the median, which represents the middle value when the data is sorted, the impact of outliers is mitigated, providing a more robust measure of the central tendency for the yearly charges.

In [12]:
# Create a dictionary containing patient information and print it
final_dict = paciente.createDictionary()

Create Dictionary:
The createDictionary method creates a structured dictionary (final_dict) containing various patient attributes. This organized data structure facilitates easy retrieval and manipulation of patient information for further analysis or reporting.

In [13]:
# Count the number of smokers in each region and print the results
print(paciente.smokers_by_region())

{'southwest': 58, 'southeast': 91, 'northeast': 67, 'northwest': 58}


Count of Smokers in Each Region:

Examining the count of smokers across regions reveals disparities, with the Southeast having the highest count (91). Understanding smoking prevalence in different regions is valuable for public health planning and targeted anti-smoking campaigns.